File: src/services/audit/detectors/ad/accounts/privileged/not-in-protected-users.ts
/**
 * Not in Protected Users Detector
 * Check for privileged accounts not in Protected Users group
 */

import { ADUser } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedUserEntities } from '../../../../../../utils/entity-converter';

export function detectNotInProtectedUsers(users: ADUser[], includeDetails: boolean): Finding {
  const privilegedGroups = ['Domain Admins', 'Enterprise Admins', 'Schema Admins'];

  const affected = users.filter((u) => {
    if (!u.memberOf) return false;
    const isPrivileged = u.memberOf.some((dn) =>
      privilegedGroups.some((group) => dn.includes(`CN=${group}`))
    );
    const isInProtectedUsers = u.memberOf.some((dn) => dn.includes('CN=Protected Users'));
    return isPrivileged && !isInProtectedUsers;
  });

  return {
    type: 'NOT_IN_PROTECTED_USERS',
    severity: 'high',
    category: 'accounts',
    title: 'Not in Protected Users Group',
    description: 'Privileged accounts not in Protected Users group. Missing additional security protections.',
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedUserEntities(affected) : undefined,
  };
}

File: internal/audit/detectors/ad/accounts/privileged/not-in-protected-users.go
package privileged

import (
	"context"
	"strings"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// NotInProtectedUsersDetector detects privileged accounts not in Protected Users group
type NotInProtectedUsersDetector struct {
	audit.BaseDetector
}

// NewNotInProtectedUsersDetector creates a new detector
func NewNotInProtectedUsersDetector() *NotInProtectedUsersDetector {
	return &NotInProtectedUsersDetector{
		BaseDetector: audit.NewBaseDetector("NOT_IN_PROTECTED_USERS", audit.CategoryAccounts),
	}
}

// Detect executes the detection
func (d *NotInProtectedUsersDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []types.User

	privilegedGroups := []string{"Domain Admins", "Enterprise Admins", "Schema Admins"}

	for _, u := range data.Users {
		if len(u.MemberOf) == 0 {
			continue
		}

		// Check if privileged
		isPrivileged := false
		for _, dn := range u.MemberOf {
			for _, group := range privilegedGroups {
				if strings.Contains(dn, "CN="+group) {
					isPrivileged = true
					break
				}
			}
			if isPrivileged {
				break
			}
		}

		if !isPrivileged {
			continue
		}

		// Check if in Protected Users
		isInProtectedUsers := false
		for _, dn := range u.MemberOf {
			if strings.Contains(dn, "CN=Protected Users") {
				isInProtectedUsers = true
				break
			}
		}

		if !isInProtectedUsers {
			affected = append(affected, u)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "Not in Protected Users Group",
		Description: "Privileged accounts not in Protected Users group. Missing additional security protections.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedUserEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewNotInProtectedUsersDetector())
}
